### PR TITLE
Fix cluster frame display issue

### PR DESCRIPTION
**Pull request by Nokel81**
- Add some defensive code to prevent this sort of infinite loop

- Add some unit tests

Signed-off-by: Sebastian Malton <sebastian@malton.name>

This is the version of https://github.com/lensapp/lens/pull/5488 for `master`. It includes some unit tests.

### CODE REVIEW

File: src/common/__tests__/cluster-store.test.ts [hunk 1]
```diff
@@ -94,6 +95,7 @@ describe("cluster-store", () => {
     mainDi.permitSideEffects(getConfigurationFileModelInjectable);
     mainDi.permitSideEffects(appVersionInjectable);
     mainDi.permitSideEffects(clusterStoreInjectable);
+    mainDi.permitSideEffects(fsInjectable);
```
Comment (jansav) [inline src/common/__tests__/cluster-store.test.ts:98]: Doesn't seem right, but neither does `mainDi.permitSideEffects(getConfigurationFileModelInjectable);`. Guess there is good reason for it for now :)

File: src/common/cluster-store/cluster-store.ts [hunk 2]
```diff
@@ -111,12 +115,24 @@ export class ClusterStore extends BaseStore<ClusterStoreModel> {
     return undefined;
   }
 
+  private createNewCluster(model: ClusterModel): Cluster {
```
Comment (jansav) [inline src/common/cluster-store/cluster-store.ts]: Not welcome here at all. This logic belongs to `createCluster` dependency function.

File: src/main/catalog-sources/__test__/kubeconfig-sync.test.ts [hunk 3]
```diff
@@ -54,10 +55,9 @@ describe("kubeconfig-sync.source tests", () => {
     di.override(kubectlDownloadingNormalizedArchInjectable, () => "amd64");
     di.override(normalizedPlatformInjectable, () => "darwin");
 
-    di.override(clusterStoreInjectable, () =>
-      ClusterStore.createInstance({ createCluster: () => null as never }),
-    );
-
+    di.permitSideEffects(fsInjectable);
```
Comment (jansav) [inline src/main/catalog-sources/__test__/kubeconfig-sync.test.ts:58]: Doesn't seem right to introduce real `fs` to unit tests.

File: src/renderer/frames/cluster-frame/cluster-frame.tsx [hunk 4]
```diff
@@ -62,7 +68,16 @@ export const NonInjectedClusterFrame = observer(({
         {
           Component
             ? <Component />
-            : <Redirect to={startUrl.get()} />
+            // NOTE: this check is to prevent an infinite loop
```
Comment (jansav) [inline src/renderer/frames/cluster-frame/cluster-frame.tsx:71]: I don't follow. How could anyone encounter this?

File: src/renderer/frames/cluster-frame/start-url.injectable.ts [hunk 5]
```diff
@@ -4,29 +4,30 @@
...
+      }
+
+      if (workloadOverviewRoute.isEnabled.get()) {
+        return workloadOverviewRoute.path;
+      }
 
-      return resources.every((resourceName) => isAllowedResource(resourceName))
-        ? clusterOverviewUrl
-        : workloadOverviewUrl;
+      /**
```
Comment (jansav) [inline src/renderer/frames/cluster-frame/start-url.injectable.ts:26]: I don't follow. Why is this needed?

File: src/common/__tests__/cluster-store.test.ts (hunk 1 above)
Comment (Nokel81) [reply src/common/__tests__/cluster-store.test.ts:98]: Well until we factor out all of this (and probably the library that we are using) we have to continue to use `mockFs`.

File: src/main/catalog-sources/__test__/kubeconfig-sync.test.ts (hunk 3 above)
Comment (Nokel81) [reply src/main/catalog-sources/__test__/kubeconfig-sync.test.ts:58]: I am not, this test previously (and still does) use `mockFs`. I just changed how it is used.

File: src/renderer/frames/cluster-frame/start-url.injectable.ts (hunk 5 above)
Comment (Nokel81) [reply src/renderer/frames/cluster-frame/start-url.injectable.ts:26]: 1. Because `workloadOverviewRoute.isEnabled` is typed as `IComputedValue<boolean>` it *could* be `false` so this is a branch of code that _could_ be run.
2. The other option is to just not check `workloadOverviewRoute.isEnabled` at all and then always return `workloadOverviewRoute.path` but then if in the future we change the definition of `workloadOverviewRoute.isEnabled` to not always be defined we might get into the same case again.

File: src/renderer/frames/cluster-frame/cluster-frame.tsx (hunk 4 above)
Comment (Nokel81) [reply src/renderer/frames/cluster-frame/cluster-frame.tsx:71]: People did: https://github.com/lensapp/lens/issues/5462

The bug was that there was a disconnect between the `Redirect` and whether `startUrl` was valid.

Comment (jansav) [reply src/renderer/frames/cluster-frame/cluster-frame.tsx:71]: Hmm. I'd say this kludge to get rid of the problem. However, this doesn't solve the real issue, which is to figure out why this is possible and make it impossible. 

Maybe good enough for now.

**Review by jansav — APPROVED** [review]

File: src/renderer/frames/cluster-frame/cluster-frame.tsx (hunk 4 above)
Comment (Nokel81) [reply src/renderer/frames/cluster-frame/cluster-frame.tsx:71]: I did fix the problem as well. This is here to prevent the same problem from occurring again in the future.

File: src/common/cluster-types.ts [hunk 6]
```diff
@@ -67,6 +67,15 @@ export interface ClusterModel {
   labels?: Record<string, string>;
 }
 
+/**
+ * This data is retreived from the kubeconfig file before calling the cluster constructor.
+ *
+ * That is done to remove the external dependency on the construction of Cluster instances.
```
Comment (jim-docker) [inline src/common/cluster-types.ts:73]: Yay!

File: src/renderer/frames/cluster-frame/cluster-frame.test.tsx [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    di.override(directoryForUserDataInjectable, () => "/some/irrelavent/path");
+    di.override(storesAndApisCanBeCreatedInjectable, () => true);
+
+    const createCluster = di.inject(createClusterInjectable);
+
+    cluster = createCluster(
+      {
+        contextName: "my-cluster",
+        id: "123456",
+        kubeConfigPath: "/irrelavent",
```
Comment (jim-docker) [inline src/renderer/frames/cluster-frame/cluster-frame.test.tsx:54]: I know it's irrelevant but it's irrelevant :)

**Review by jim-docker — APPROVED** [review]

File: src/renderer/frames/cluster-frame/cluster-frame.test.tsx (hunk 7 above)
Comment (Nokel81) [reply src/renderer/frames/cluster-frame/cluster-frame.test.tsx:54]: LoL
